### ISSUE TITLE
JsonConvert.DeserializeObject<Stack<T>>/JsonConvert.Serialize(Stack<T) does not work as expected.

### ISSUE THREAD

**Issue by rymeskar** [JamesNK/Newtonsoft.Json#971]
When I serialize a Stack of strings, or of anything else for that matter, and then try to deserialize it with the functions mentioned in the headline, I get a stack with reversed order. Is this unexpected behavior an intended feature? Then to get the expected behavior, I would have to write a specific code for handling stack?

Code that revealed this "feature":
```csharp
var stack = new Stack<string>();
stack.Push("Inserted first, should be popped last");
stack.Push("Inserted last, should be popped first");
var serializedStack = JsonConvert.SerializeObject(stack);
Console.WriteLine(serializedStack);
//["Inserted last, should be popped first","Inserted first, should be popped last"]
stack = JsonConvert.DeserializeObject<Stack<string>>(serializedStack);
serializedStack = JsonConvert.SerializeObject(stack);
Console.WriteLine(serializedStack);
//["Inserted first, should be popped last","Inserted last, should be popped first"]
Console.WriteLine(stack.Pop());
//Inserted first, should be popped last
Console.WriteLine(stack.Pop());
//Inserted last, should be popped first
 ```

**Comment by JamesNK** [comment]
I'm afraid this is a limition of a Stack. The results returned when it is serialized and the opposite order for when it is deserialized.

**Comment by SlimeQ** [comment]
How could this possibly be a limitation of the stack? The library should just convert to a list and reverse it before serializing. Or when deserializing, deserialize first to a list, reverse it, and create a stack from that list

**Comment by Banyc** [comment]
This issue has not been fixed even in version 12.0.3. Sad.
